### PR TITLE
Don't block etcd client

### PR DESCRIPTION
**Pull request by lavalamp**
@bprashanth You've been complaining about this channel for ages. What do you think about this?

### CODE REVIEW

**Comment by bprashanth** [comment]
I think we can spare an MB to the watchers. Though I suspect this is expiration at the go-etcd request watch level, or a combination of the two (i.e I could reproduce https://github.com/GoogleCloudPlatform/kubernetes/issues/8700 reliably without apiserver). I also think we can try bumping up the etcd watch window itself, from the little digging ive don't I don't think it should hog that much memory. 

I'm still puzzled what's writing so much. 

Eitherway this change should be safe since you're just buffering a channel and we don't have *that* many concurrent watches that we oom because of this (we have ~500 in a 100 node cluster i think).

**Comment by bprashanth** [comment]
Shippable looks like their go install failure. I guess you need more lgtms to merge.

**Comment by alex-mohr** [comment]
Candidate for 20.1?  #10409

**Comment by alex-mohr** [comment]
@dbsmith and this is a fix (or mitigation) to #10412?

**Comment by bprashanth** [comment]
This isn't a fix for https://github.com/GoogleCloudPlatform/kubernetes/issues/10412, it's a guess at what might ameliorate our recently observed performance regression

**Comment by davidopp** [comment]
LGTM

**Comment by davidopp** [comment]
@dchen1107 @quinton-hoole for more LGTMs

**Comment by wojtek-t** [comment]
Thanks @lavalamp - LGTM

@dchen1107 @quinton-hoole - can one of you also LGTM it?

**Comment by dchen1107** [comment]
ok to test

LGTM

File: pkg/tools/etcd_helper_watch.go [hunk 1]
```diff
@@ -250,6 +259,10 @@ func (w *etcdWatcher) translate() {
 			return
 		case res, ok := <-w.etcdIncoming:
 			if ok {
+				if curLen := len(w.etcdIncoming); curLen > 0 {
```
Comment (quinton-hoole) [inline pkg/tools/etcd_helper_watch.go]: This is going to cause a logging storm, which will further exacerbate the backlog. Don't you rather want this logging threshold to be greater than zero?  Perhaps n/2 for a channel of length n?  And you don't want to log more often than once a minute or so.

Comment (lavalamp) [inline pkg/tools/etcd_helper_watch.go]: Basically I want to know if it's ever > 0-- if so, that justifies this patch. If not, this patch didn't fix anything.

I can adjust to reduce log spam, but the decoding process that is backing things up is *much* more resource intensive than a log message, so this is "only" a cosmetic problem.

Comment (quinton-hoole) [inline pkg/tools/etcd_helper_watch.go]: I'd suggest basic edge triggering then.  Log once when it crosses from zero to non-zero, and once when it crosses back again.

Comment (wojtek-t) [inline pkg/tools/etcd_helper_watch.go]: @quinton-hoole - this is logged at the 4 level - by default we're using 2, so this won't be logged

Comment (quinton-hoole) [inline pkg/tools/etcd_helper_watch.go]: Good point.  Fair enough. LGTM

**Comment by wojtek-t** [comment]
IIUC, we have now 3 lead-lgtms so it's ready to merge, right?

**Comment by quinton-hoole** [comment]
Yup.

**Comment by alex-mohr** [comment]
Shippable still failing and it's not merged.  I need to cut v0.20.1 shortly -- plan?

**Pull request merged**